### DISCUSSION TITLE
Add non-cryptographic hashing features such as xxhash and blake3 for cryptographic

### DISCUSSION THREAD

**Discussion by kikyps** [jedisct1/libsodium#1316] [Ideas]
Add a super fast Hashing feature for non-cryptography using the algorithm from xxhash
and is there any idea to upgrade blake2b to latest and faster blake3?

thank you very much for your hard work

**Comment by jedisct1** [comment]
This is a crypto library, so no, there are no plans to add non-cryptographic hash functions.
For a non-cryptographic hash, take a look at [polymur hash](https://github.com/orlp/polymur-hash).

BLAKE3 has a smaller security margin than BLAKE2, and is hostile to implementers and CPU caches. TurboSHAKE/KangarooTwelve are better alternatives, and are more likely to be added as XOFs. They are about as fast, much cleaner, and are being standardized.

If you have a secret key, need something super fast and only need a 256 bit output, you can use `crypto_aead_aegis128l()` today.

**Reply by kikyps** [reply]
Thank you for the explanation, what library do you think is powerful and fast enough to check file integrity? so far what I know and now I use is xxhash and until now I am still looking to learn and compare which ones are accurate and fast, and will TurboSHAKE/KangarooTwelve be added in the future will it replace blake2 or will both still be used?

**Reply by jedisct1** [reply]
Depends on who is creating the files and who will verify them. Is it the same application?

`crypto_generichash()` will still be BLAKE2.

**Reply by kikyps** [reply]
No I just want to do a file check inside my application to detect tampering.

Is there a way to protect openat from hooking/intercepting, which can hook or intercept the openat path by changing the original path to a specified path, can the sodium_mprotect and sodium_mlock functions handle this case to protect the memory area so that openat cannot be hooked so that it always leads to the original path, maybe in the future there will be sodium_openat for this case? XD

I am interested in the new features in the last update, I will learn more about AEADs.

**Reply by jedisct1** [reply]
If this is inside your own application, you'd better use a MAC, not a hash function (i.e. use a secret key).

Create a 16 bytes secret key with `crypto_aead_aegis128l_keygen()`, and then use `crypto_aead_aegis128l_encrypt_detached(NULL, mac, NULL, NULL, 0, file_content, file_length, NULL, nonce, key)`.

Note that this doesn't encrypt anything, but it authenticates `file_content` (which is the associated data).

You'd get a 32 byte authenticator in `mac`.

The `nonce` can be set to all zeros, or to 16 arbitrary bytes that represent your application or identifies the file.

That's the fastest option.

**Reply by kikyps** [reply]
How does it work?
does it work anyway comparing mac with expected mac?
is there a complete documentation reference about this function?
this is very interesting.

Thanks

**Reply by jedisct1** [reply]
https://doc.libsodium.org/secret-key_cryptography/aead/aegis-128l

Just fill the additional data, keep the message empty.

But first check if `generichash` is really not fast enough. For the vast majority of application, it likely is.
